File: eco-core/core-nms/v1_17_R1/src/main/java/com/willfp/eco/proxy/v1_17_R1/VillagerTrade.java
package com.willfp.eco.proxy.v1_17_R1;

import com.willfp.eco.core.display.Display;
import com.willfp.eco.proxy.VillagerTradeProxy;
import net.minecraft.world.item.trading.MerchantOffer;
import org.bukkit.craftbukkit.v1_17_R1.inventory.CraftMerchantRecipe;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.MerchantRecipe;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Field;

public final class VillagerTrade implements VillagerTradeProxy {
    private final Field handle;

    public VillagerTrade() {
        try {
            handle = CraftMerchantRecipe.class.getDeclaredField("handle");
            handle.setAccessible(true);
            return;
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        }

        throw new RuntimeException("Error!");
    }

    @Override
    public MerchantRecipe displayTrade(@NotNull final MerchantRecipe recipe,
                                       @NotNull final Player player) {
        CraftMerchantRecipe oldRecipe = (CraftMerchantRecipe) recipe;

        CraftMerchantRecipe newRecipe = new CraftMerchantRecipe(
                Display.display(recipe.getResult().clone()),
                recipe.getUses(),
                recipe.getMaxUses(),
                recipe.hasExperienceReward(),
                recipe.getVillagerExperience(),
                recipe.getPriceMultiplier()
        );

        for (ItemStack ingredient : recipe.getIngredients()) {
            newRecipe.addIngredient(Display.display(ingredient.clone(), player));
        }

        getHandle(newRecipe).setSpecialPriceDiff(getHandle(oldRecipe).getSpecialPriceDiff());

        return newRecipe;
    }

    @NotNull
    private MerchantOffer getHandle(@NotNull final CraftMerchantRecipe recipe) {
        try {
            return (MerchantOffer) handle.get(recipe);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }

        throw new IllegalArgumentException("Not CMR");
    }
}

File: eco-core/core-nms/v1_17_R1/src/main/kotlin/com/willfp/eco/proxy/v1_17_R1/VillagerTrade.kt
package com.willfp.eco.proxy.v1_17_R1

import com.willfp.eco.core.display.Display
import com.willfp.eco.proxy.VillagerTradeProxy
import net.minecraft.world.item.trading.MerchantOffer
import org.bukkit.craftbukkit.v1_17_R1.inventory.CraftMerchantRecipe
import org.bukkit.entity.Player
import org.bukkit.inventory.MerchantRecipe
import java.lang.reflect.Field

class VillagerTrade : VillagerTradeProxy {
    private lateinit var handle: Field

    override fun displayTrade(
        recipe: MerchantRecipe,
        player: Player
    ): MerchantRecipe {
        val oldRecipe = recipe as CraftMerchantRecipe
        val newRecipe = CraftMerchantRecipe(
            Display.display(recipe.getResult().clone()),
            recipe.getUses(),
            recipe.getMaxUses(),
            recipe.hasExperienceReward(),
            recipe.getVillagerExperience(),
            recipe.getPriceMultiplier()
        )
        for (ingredient in recipe.getIngredients()) {
            newRecipe.addIngredient(Display.display(ingredient.clone(), player))
        }
        getHandle(newRecipe).specialPriceDiff = getHandle(oldRecipe).specialPriceDiff
        return newRecipe
    }

    private fun getHandle(recipe: CraftMerchantRecipe): MerchantOffer {
        try {
            return handle[recipe] as MerchantOffer
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
        throw IllegalArgumentException("Not CMR")
    }

    init {
        try {
            handle = CraftMerchantRecipe::class.java.getDeclaredField("handle")
            handle.isAccessible = true
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        }
        throw RuntimeException("Error!")
    }
}